src/app/view/device/add-devices/add-devices.component.ts

```diff
@@ -362,28 +396,168 @@ export class AddDevicesComponent {
   }
+  checkDocumentsUploaded() {
+    if (Object.keys(this.files).length === 0) {
+      this.allDocumentsUploaded = false;
+      return;
+    }
+    this.allDocumentsUploaded = this.deviceForms.controls.every(
+      (_, deviceIndex) => {
+        if (!this.files[deviceIndex]) return false;
+
+        return this.requiredFileTypes.every(
+          (fileType) =>
+            this.files[deviceIndex][fileType] &&
+            this.files[deviceIndex][fileType].length > 0,
+        );
+      },
+    );
+
+    this.checkFormValidity();
+  }
+  checkFormValidity() {
+    const formValid = this.myform.valid;
+
+    const docsValid = this.deviceForms.controls.every((_, deviceIndex) => {
+      if (!this.files[deviceIndex]) return false;
+
+      return this.requiredFileTypes.every(
+        (fileType) =>
+          this.files[deviceIndex][fileType] &&
+          this.files[deviceIndex][fileType].length > 0,
+      );
+    });
+
+    this.formValid = formValid && docsValid;
+  }
+  onFileChange(event: Event, deviceIndex: number, fileType: FileType) {
+    const input = event.target as HTMLInputElement;
+    if (!input.files || input.files.length === 0) return;
+
+    const files: FileList = input.files;
+
+    if (!this.files[deviceIndex]) {
+      this.files[deviceIndex] = {
+        productionFacilityRegistration: [],
+        ownershipProof: [],
+        meteringEvidence: [],
+        singleLineDiagram: [],
+        projectPhotos: [],
+      };
+    }
+
+    this.files[deviceIndex][fileType] = Array.from(files);
+
+    const fileControl = this.deviceForms.at(deviceIndex).get(fileType);
+    if (fileControl) {
+      fileControl.setValue(input.files[0]);
+      fileControl.markAsDirty();
+    }
+    this.checkDocumentsUploaded();
+  }
 
   onSubmit() {
     if (this.myform.valid) {
       this.openPopupDialog();
+      this.isSubmitting = true;
     }
   }
 
   submitForm() {
     const deviceArray = this.myform.value.devices;
-    deviceArray.forEach((element: any) => {
+    deviceArray.forEach((element: any, index: number) => {
+      const formData = new FormData();
+
       if (this.organizationName != null) {
         element['organizationId'] = this.organizationId;
       }
       const selectedCountry = this.countrylist.find(
         (option: CountryInfo) => option.country === element.countryCodename,
       );
       element['countryCode'] = selectedCountry?.alpha3;
-      this.deviceService.Postdevices(element).subscribe({
+      formData.append('deviceToRegister', JSON.stringify(element));
+      if (element.countryCode) {
+        formData.append('countryCode', element.countryCode);
+      } else {
+        console.error('Country code is missing for device:', element);
+      }
+
+      const fileFields: FileType[] = [
+        'productionFacilityRegistration',
+        'ownershipProof',
+        'meteringEvidence',
+        'singleLineDiagram',
+        'projectPhotos',
+      ];
+
+      const allowedExtensions = [
```

Comment:
can we have the document processing for upload into a reusable function

Reply:
I would suggest having these into a constant or something that can be re-used. We can have extensions for images, for documents and similar. Those formats are usually universal and can be specific based on the type of document you are uploading. (They are scoped based on the document not on the type of upload)